apps/grader-host/src/lib/jobLogger.ts

```diff
@@ -0,0 +1,56 @@
+import winston = require('winston');
+import Transport = require('winston-transport');
+import { Writable, type WritableOptions } from 'node:stream';
+
+import { config } from './config';
+
+interface WinstonBufferedLogger extends winston.Logger {
+  getBuffer(): Buffer;
+}
+
+export class BufferedWritableStream extends Writable {
+  buffer: any[];
+
+  constructor(options?: WritableOptions) {
+    super(options);
+    this.buffer = [];
+  }
+
+  _write(chunk: any, _encoding: BufferEncoding, callback: (error?: Error | null) => void) {
+    this.buffer.push(chunk);
```

Comment:
Can we add a limit here for the total log size (maybe 1 MB?) so that if a grader goes crazy we won't exhaust memory (and S3) buffering logs?

Reply:
Great idea! I'll add that.

Reply:
Added in 564cda3. I'll test on staging once more before merging this.